lib/ProductOpener/Ingredients.pm

```diff
@@ -2831,8 +2831,16 @@ sub analyze_ingredients($) {
 					$property_value = "en:may-contain-" . $from_what_with_dashes ; # en:may-contain-palm-oil
 					$ingredients_analysis_ref->{$property_value} = $values{maybe};
 				}
-				elsif (defined $values{unknown_ingredients}) {
+				# If some ingredients are not recognized, there is a possibility that they could be palm oil or contain palm oil
+				# As there are relatively few ingredients with palm oil, we assume we are able to recognize them with the taxonomy
+				# and that unrecognized ingredients do not contain palm oil.
+				# --> We mark the product as palm oil free
+				# Exception: If there are lots of unrecognized ingredients though (e.g. more than 1 third), it may be that the ingredients list
+				# is bogus (e.g. OCR errors) and the likelyhood of missing a palm oil ingredient increases.
+				# --> In this case, we mark the product as palm oil content unknown
+				elsif ((defined $values{unknown_ingredients}) and ((scalar @{$values{unknown_ingredients}}) > (scalar(@{$product_ref->{ingredients}}) / 3))) {
```

Comment:
This condition is only for palm-oil ?

Reply:
I would really change that:
1. for readability 
2. to be more explicit
3. to keep analysis capability even if we declare palm-oil free (it will be useful, if eg. one unknown ingredient did indeed contains palm oil)

```perl
			elsif (defined $values{unknown_ingredients}) {
				# Some ingredients were not recognized
				$log->debug("analyze_ingredients - unknown ingredients", { unknown_ingredients_n => (scalar @{$values{unknown_ingredients}}), ingredients_n => (scalar(@{$product_ref->{ingredients}})) }) if $log->is_debug();
				my $unknown_rate = (scalar @{$values{unknown_ingredients}}) / (scalar(@{$product_ref->{ingredients}});
				# for palm-oil, as there are few products containing it, we consider status to be unknown only if there is more than 30% unknown ingredients (which may indicates bogus ingredient list, eg. OCR errors)
				if ($from_what_with_dashes neq "palm-oil" || $unknown_rate > .3) {
				    $property_value = "en:" . $from_what_with_dashes . "-content-unknown"; # en:palm-oil-content-unknown
				}
				$ingredients_analysis_ref->{$property_value} = $values{unknown_ingredients};
```

Reply:
good idea, I'll do something like that.